wurst/systems/spells/gatherer/TeleGathering.wurst

```diff
@@ -0,0 +1,161 @@
+    vec2 pos
+    construct(vec2 pos)
+        this.pos = pos
+
+class TeleGatherInstance
+    unit caster
+    unit teleTarget
+    real duration
+    real lifetime = 0
+    UnitEntity casterEntity
+    CallbackPeriodic pollingLoop
+    IterableMap<int, SortingPosition> sortingPositions
+    real nextSortAngle = 0
+    real sortDistanceOffset = SORT_DISTANCE_INCREMENT
+
+    construct(unit caster, unit teleTarget, real duration)
+        InstantDummyCaster.castTarget(caster.getOwner(), ABILITY_ID_CASTER_BUFF, 1, "curse", caster, caster.getPos() + vec2(100,0))
+        this.caster = caster
+        this.casterEntity = UnitEntity.findForUnit(this.caster)
+        this.duration = duration
+        this.teleTarget = teleTarget
+        this.sortingPositions = new IterableMap<int, SortingPosition>()
+        startBuffPolling()
+
+    function startBuffPolling()
+        this.pollingLoop = doPeriodically(1) cb ->
+            this.lifetime += 1
+            if this.lifetime >= this.duration
+                or not caster.hasAbility(CASTER_BUFF_ID)
+                or not teleTarget.hasAbility(TARGET_BUFF_ID)
+                destroy this
+
+    function canGatherItem(int itemType) returns bool
+        if itemType == 'XRAW' //Raw meat
+            return false
+
+        //Skip herbs if not herb-gatherer
+        if not this.casterEntity instanceof HerbMaster
+            and not this.casterEntity instanceof Omnigatherer
+                if itemType == ITEM_RIVER_ROOT
+                    or itemType == ITEM_RIVER_STEM
+                    or itemType == ITEM_BUTSU
+                    or itemType == ITEM_BLUE_HERB
+                    or itemType == ITEM_ORANGE_HERB
+                    or itemType == ITEM_YELLOW_HERB
+                    or itemType == ITEM_PURPLE_HERB
+                    or itemType == ITEM_MUSHROOM
+                    return false
+
+        return true
+
+    function handleTeleGather(item itm)
+        let itemType = itm.getTypeId()
+
+        if not canGatherItem(itemType)
+            return
+
+        let sortingPos = this.getSortingPos(itemType)
+        this.caster.removeItem(itm)
+        itm.setPos(sortingPos.pos)
+
+    function getSortingPos(int itemType) returns SortingPosition
+        let targetPos = teleTarget.getPos()
+
+        //If a sorting position exists for type, return that
+        if this.sortingPositions.has(itemType)
+            return this.sortingPositions.get(itemType)
+
+        //If a nearby item of the same type is found, set up sorting position there
+        let nearest = findNearestItem(targetPos, sortDistanceOffset * 2.2, i -> i.getTypeId() == itemType)
+        if nearest != null
+            let pos = new SortingPosition(nearest.getPos())
+            sortingPositions.put(itemType, pos)
+            return pos
+
+        //Otherwise set up sorting position in a radius around the fire
+        var maybeTargetPos = targetPos.polarOffset(this.nextSortAngle.fromDeg(), this.sortDistanceOffset)
```

Comment:
I think you can avoid duplicating logic, even though Wurst doesn't have a do-while loop. You could see this equal to a placeholder, check that in the `while` statement as well, and move the increment behavior to the end of the loop.

Reply:
Changed